app/api/v2/identity/sessions.rb

```diff
@@ -111,29 +117,111 @@ def get_user(email)
+            user = User.find_by_uid(user_uid)
+            error!({ errors: ['identity.session.not_found'] }, 404) unless user
+
+            switch_oid = params[:oid]
+            # Check switch session mode. Switch proceed only if params[:oid] provided
+            unless switch_oid.nil?
+              # Verify barong admin has AdminSwitchSession permission
+              if admin_organization? :read, AdminSwitchSession
+                # User is barong organization admin
+                oids = Organization.all.pluck(:id)
+              else
+                # User is organization admin/account
+                # Check account in the organization that user belong to
+                members = Membership.joins('LEFT JOIN organizations ON organizations.id = memberships.organization_id')
+                                    .where(user_id: user.id)
+                                    .select('memberships.*,organizations.name, organizations.parent_id')
+                                    .pluck(:organization_id, :'organizations.name', :'organizations.parent_id')
+                                    .map { |id, name, pid| { id: id, name: name, pid: pid } }
+                error!({ errors: ['identity.member.not_found'] }, 404) if members.nil? || members.length.zero?
+
+                oids = Organization.where(id: members.pluck(:id)).pluck(:id)
+                members.select { |m| m[:pid].nil? }.each do |m|
+                  oids.concat(Organization.where(parent_id: m[:id]).pluck(:id))
+                end
+              end
+              error!({ errors: ['identity.member.not_found'] }, 404) if oids.length.zero?
+
+              org = Organization.find_by_oid(switch_oid)
+              error!({ errors: ['identity.member.not_found'] }, 404) if org.nil?
+              error!({ errors: ['identity.member.not_found'] }, 404) unless oids.include? org.id
+
+              if org.parent_id.nil?
+                # User switch as organization admin
+                organization_oid = switch_oid
+              else
+                # User switch as organization subunit
+                organization = Organization.find(org.parent_id)
+                error!({ errors: ['identity.organization.not_found'] }, 404) unless organization
+
+                organization_oid = organization.oid
+              end
+
+              switch = {
+                uid: switch_oid,
+                oid: organization_oid,
+                rid: user.uid
+              }
             end
 
-            activity_record(user: user.id, action: 'login', result: 'succeed', topic: 'session')
-            csrf_token = open_session(user)
-            publish_session_create(user)
+            activity_record(user: user.id, action: (switch_oid.nil? ? 'switch::user' : 'switch::orgnization'),
+                            result: 'succeed', topic: 'session')
+            Barong::RedisSession.delete(user.uid, session.id)
+            session.destroy
+
+            if switch.nil?
```

Comment:
I don't understand what we are doing here, we should stop the execution before if something is wrong in params.

Reply:
This is for checking the swith mode. You can use the same endpoint to be switch or destroy switch mode.
It's depend on the calculation of `switch` from above process.